libcontainer/cgroups/fs/rdma.go

```diff
@@ -0,0 +1,118 @@
+// +build linux
+
+package fs
+
+import (
+	"github.com/opencontainers/runc/libcontainer/cgroups"
+	"github.com/opencontainers/runc/libcontainer/cgroups/fscommon"
+	"github.com/opencontainers/runc/libcontainer/configs"
+	"math"
+	"strconv"
+	"strings"
+)
+
+type RdmaGroup struct {
+}
+
+func (s *RdmaGroup) Name() string {
+	return "rdma"
+}
+
+func (s *RdmaGroup) Apply(path string, d *cgroupData) error {
+	return join(path, d.pid)
+}
+
+func createCmdString(device string, limits configs.LinuxRdma) string {
+	var cmdString string
+	cmdString = device
+	if limits.HcaHandles != nil {
+		cmdString = cmdString + " " + "hca_handle=" + strconv.FormatUint(uint64(*limits.HcaHandles), 10)
+	}
+	if limits.HcaObjects != nil {
+		cmdString = cmdString + " " + "hca_object=" + strconv.FormatUint(uint64(*limits.HcaObjects), 10)
+	}
+	return cmdString
+}
+
+func (s *RdmaGroup) Set(path string, cgroup *configs.Cgroup) error {
+	for device, limits := range cgroup.Resources.Rdma {
+		if err := fscommon.WriteFile(path, "rdma.max", createCmdString(device, limits)); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *RdmaGroup) GetStats(path string, stats *cgroups.Stats) error {
+	if !cgroups.PathExists(path) {
+		return nil
+	}
+	currentData, err := fscommon.GetCgroupParamString(path, "rdma.current")
+	if err != nil {
+		return err
+	}
+	currentPerDevices := strings.Split(string(currentData), "\n")
+	maxData, err := fscommon.GetCgroupParamString(path, "rdma.max")
+	if err != nil {
+		return err
+	}
+	maxPerDevices := strings.Split(string(maxData), "\n")
+	// If device got removed between reading two files, ignore returning
+	// stats.
+	if len(currentPerDevices) != len(maxPerDevices) {
+		return nil
+	}
+	currentEntries := ConvertRdmaEntry(currentPerDevices)
+	maxEntries := ConvertRdmaEntry(maxPerDevices)
+
+	stats.RdmaStats = cgroups.RdmaStats{
+		RdmaLimit:   maxEntries,
+		RdmaCurrent: currentEntries,
+	}
+
+	return nil
+}
+
+// Parse raw string to RdmaEntry
+func ParseRdmaKV(raw string, entry *cgroups.RdmaEntry) {
```

Comment:
Why this function is public?

Reply:
@kolyshkin Following function is intentionally public as being used by rdma controller in `fs2`. I didnt wanted to have redundant code in both implementations of `fs` and `fs2` and these functions are exclusive to Rdma so i thought i'd rather keep them with rdma and not move to `fscommon`

Reply:
fscommon can hold the helper for RDMA. fs2 imports fscommon better than fs package. just my two cents.

Reply:
@fuweid Oh so idea was since `ParseRdmaKV` and `ConvertRdmaEntry` uses types from package `cgroups` having them in `fscommon` will introduce `import cycles`. Since both of these functions are only going to be used only by `rdma` therefore i thought its better to keep them with `rdma` implementation only. But let me know if you guys think otherwise we can figure out a way. :smile:

Reply:
I prefer to keep the duplicate code in separate packages (non-binding). :)

Reply:
I think this should go to `fscommon`, which naturally contains the stuff that's common between fs and fs2.

Reply:
Which is a problem, as described above (https://github.com/opencontainers/runc/pull/2883/commits/02eb7236c86c4b700c7465cf4779ae12bf1a19a3#r610598989).

To solve it, I think we should move OpenFile/ReadFile etc from fscommon to `cgroups/utils`.

Reply:
https://github.com/opencontainers/runc/pull/3024

Reply:
Resolved in latest commit.